cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRequestQueue.java

```diff
@@ -0,0 +1,128 @@
+import io.cdap.cdap.app.preview.PreviewRequest;
+import io.cdap.cdap.app.preview.PreviewRequestQueue;
+import io.cdap.cdap.app.preview.PreviewRequestQueueState;
+import io.cdap.cdap.app.store.preview.PreviewStore;
+import io.cdap.cdap.common.ConflictException;
+import io.cdap.cdap.common.app.RunIds;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.proto.id.ApplicationId;
+import net.jcip.annotations.ThreadSafe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Thread-safe implementation of {@link PreviewRequestQueue} backed by {@link PreviewStore}.
+ */
+@ThreadSafe
+public class DefaultPreviewRequestQueue implements PreviewRequestQueue {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultPreviewRequestQueue.class);
+  private final PreviewStore previewStore;
+  private final ConcurrentLinkedDeque<PreviewRequest> requestQueue;
+  private final int capacity;
+  private final long waitTimeOut;
+
+  @Inject
+  public DefaultPreviewRequestQueue(CConfiguration cConf, PreviewStore previewStore) {
+    this.previewStore = previewStore;
+    this.capacity = cConf.getInt(Constants.Preview.WAITING_QUEUE_CAPACITY, 50);
+    this.waitTimeOut = cConf.getLong(Constants.Preview.WAITING_QUEUE_TIMEOUT_SECONDS, 60);
+    this.requestQueue = new ConcurrentLinkedDeque<>();
+    List<PreviewRequest> allInWaitingState = previewStore.getAllInWaitingState();
+    for (PreviewRequest request : allInWaitingState) {
+      if (isTimedOut(request)) {
+        continue;
+      }
+      requestQueue.add(request);
+    }
+  }
+
+  @Override
+  public Optional<PreviewRequest> poll(byte[] pollerInfo) {
+    while (true) {
+      PreviewRequest previewRequest = requestQueue.poll();
+      if (previewRequest == null) {
+        return Optional.empty();
+      }
+
+      if (isTimedOut(previewRequest)) {
+        LOG.warn("Preview request wth application id {} is timed out. Ignoring it.",
+                 previewRequest.getProgram().getParent());
+        continue;
+      }
+
+      try {
+        previewStore.setPreviewRequestPollerInfo(previewRequest.getProgram().getParent(), pollerInfo);
+      } catch (ConflictException e) {
+        LOG.debug("Preview application with id {} is not present in WAITING state. Ignoring the preview.",
+                  previewRequest.getProgram().getParent());
+        continue;
+      } catch (Exception e) {
+        LOG.warn("Error while setting the poller info for preview request with application id {}. Trying again",
+                 previewRequest.getProgram().getParent(), e);
+        requestQueue.addFirst(previewRequest);
+        continue;
+      }
+
+      return Optional.of(previewRequest);
+    }
+  }
+
+  @Override
+  public synchronized void add(PreviewRequest previewRequest) {
```

Comment:
Shouldn't synchronize the whole method. You can use a `BlockingDeque` to limit the capacity so that it can never grow beyond a given size in a concurrent safe way.

Reply:
done.